src/device/device_ledger.cpp

```diff
@@ -532,15 +587,15 @@ namespace hw {
     /* ======================================================================= */
 
     bool device_ledger::derive_subaddress_public_key(const crypto::public_key &pub, const crypto::key_derivation &derivation, const std::size_t output_index, crypto::public_key &derived_pub){
-
-      lock_device();
-      try {
-        int offset =0;
-        unsigned char options = 0;
-
+        AUTO_LOCK_CMD();
         #ifdef DEBUG_HWDEVICE
         const crypto::public_key pub_x = pub;
-        const crypto::key_derivation derivation_x  = hw::ledger::decrypt(derivation);
+        crypto::key_derivation derivation_x;
+         if ((this->mode == TRANSACTION_PARSE) && has_view_key) {    
```

Comment:
This check `mode == TRANSACTION_PARSE ` seems wrong, because `has_view_key` being `true` already implies that `derivation` is unencrypted. I think the current code will cause an error when it's called with `mode` being `TRANSACTION_CREATE_REAL` and `has_view_key` being `true`. Have you tested this code?

Reply:
Now I realized that this function `derive_subaddress_public_key()` is only called when scanning for incoming txes with `is_out_to_acc_precomp`, so I still think my above statement valid. In other words, this code is never executed when `this->mode != TRANSACTION_PARSE`, so the expression `(this->mode == TRANSACTION_PARSE) && has_view_key` is the same as `has_view_key`. Not that the current code is broken or anything, but it adds some unnecessary confusion.

Reply:
Seems like I was misunderstanding.